KAFKA-6726: Fine Grained ACL for CreateTopics (KIP-277)

edoardocomar opened the pull request:
* Handling CreateTopicsRequest now requires Create auth on Topic resource and does not require Create on Cluster

* AclCommand --producer option adjusted 

* Existing Unit and Integration tests adjusted accordingly

https://issues.apache.org/jira/browse/KAFKA-6726
https://cwiki.apache.org/confluence/display/KAFKA/KIP-277+-+Fine+Grained+ACL+for+CreateTopics+API

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mimaison committed:
KAFKA-6726: KIP-277 - Fine Grained ACL for CreateTopics API

* Handling CreateTopicsRequest now requires Create auth on Topic
resource and does not require Create on Cluster

* AclCommand --producer option adjusted

* Existing Unit and Integration tests adjusted accordingly

https://issues.apache.org/jira/browse/KAFKA-6726
https://cwiki.apache.org/confluence/display/KAFKA/KIP-277+-+Fine+Grained+ACL+for+CreateTopics+API

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

- [ ] Verify design and implementation
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma commented:
Thanks for the PR, let's add an upgrade note, please.

edoardocomar commented:
Hi @ijuma the PR needs to be updated to match the KIP amends

edoardocomar committed:
upgrade notes

edoardocomar committed:
minor improvement to test with auto-create option

edoardocomar commented:
Hi @ijuma thanks for reviewing
we rebased, marginally improved testing for auto-create and added upgrade notes done

edoardocomar committed:
KAFKA-6726: KIP-277 - Fine Grained ACL for CreateTopics API

* Handling CreateTopicsRequest now requires Create auth on Topic
resource and does not require Create on Cluster

* AclCommand --producer option adjusted

* Existing Unit and Integration tests adjusted accordingly

* upgrade notes

https://issues.apache.org/jira/browse/KAFKA-6726
https://cwiki.apache.org/confluence/display/KAFKA/KIP-277+-+Fine+Grained+ACL+for+CreateTopics+API

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

- [ ] Verify design and implementation
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

edoardocomar commented:
@ijuma last commit was just to resolve merge conflicts

edoardocomar commented:
test failure unrelated (dynamic broker config)

@ijuma can you please review as the deadline for feature freeze is tomorrow

ijuma commented:
This is a small feature, so there's an extra week.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -815,8 +848,9 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
-    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Create)), Resource.ClusterResource)
+    // using topic-level create Acl in this test, leaving cluster-level create Acl in the matching `write` test
+    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write), new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Create)),
+      newTopicResource)
```
It's safer to have asserts for both cases here and in the other place.

edoardocomar replied:
Ok, added test cases for all code paths

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -223,16 +235,45 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
+  }
+
+  @Test
+  def testProduceConsumeTopicAutoCreateClusterCreateAcl(): Unit = {
```
Do we need new tests here given `AuthorizerIntegrationTest`.

edoardocomar replied:
yes, we think so because EndToEndAuthorizationTest uses the acl tool command line

ijuma replied:
Don't we have `AclCommandTest` for that?

ijuma replied:
Generally, this test was meant to cover some high level end to end cases. It seems OK to add an auto create case, but overkill to add one for Create Topic and another for Create Cluster. We should add a test to `AclCommandTest` for the latter.

edoardocomar replied:
Thanks, I see the `AclCommandTest` has already coverage of adding cluster create and topic create acls. The `AuthorizerIntegrationTest` covers all request handling logic with and without auto-create.

So we only left one end2end test case for auto create with topic create acl

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -491,6 +493,37 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+
+    val requestKeysToAcls = Map[ApiKeys, Map[Resource, Set[Acl]]](
+      ApiKeys.CREATE_TOPICS -> clusterCreateAcl,
+    )
```
It's a bit odd that we are using maps here even though there's a single entry in each. Shall we simplify?

edoardocomar replied:
agreed!

ijuma reviewed (commented):
Thanks for the PR, a few minor comments after the first pass.

edoardocomar committed:
update following review

check all paths on topic auto creation

simplified testCreateTopicAuthorizationWithClusterCreate

edoardocomar commented:
Hi @ijuma thanks for the review - we rebased to address the conflicts and also addressed at your comments 

you will notice a weird double call to `consumers.head.poll(Duration.ofMillis(50L));`
as we wanted the test to check topic auto-creation without a write request but for some reason (unrelated to our changes) the first call to `poll` after a previously failed one did not trigger the topic auto-creation.

edoardocomar committed:
avoiding double poll in test

edoardocomar commented:
about the need for a different poll timeout - I opened another jira, does not seem a critical bug 
https://issues.apache.org/jira/browse/KAFKA-6994

ijuma commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -153,13 +153,16 @@ object AclCommand extends Logging {
-      (Resource.ClusterResource -> (getAcl(opts, Set(Create)) ++
-        (if (enableIdempotence) getAcl(opts, Set(IdempotentWrite)) else Set.empty[Acl])))
+    //Write, Describe, Create permission on topics, Write, Describe on transactionalIds
+    topics.map(_ -> topicAcls).toMap[Resource, Set[Acl]] ++
```
Can `toMap[Resource, Set[Acl]]` be simply `toMap`? Same for other cases.

edoardocomar replied:
thanks, existing syntax simplified

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -800,27 +839,39 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
-      this.consumers.head.assign(List(topicPartition).asJava)
-      consumeRecords(this.consumers.head)
-      Assert.fail("should have thrown exception")
+      this.consumers.head.poll(Duration.ofMillis(50L));
```
Why did we replace `consumeRecords` with `poll`?

edoardocomar replied:
We modified the method to avoid producing any records - to avoid adding write permission which made the test more complicated than it needed to be.

So we need to pass `numRecords=0` and we had no control on the timeout, unless we made that another optional argument.
without actually consuming records it seems to me that just calling `poll` is simpler than `consumeRecords`

ijuma replied:
We want to be generous with the timeout then. Any reason why it's so low?

edoardocomar replied:
the default timeout was fine for the first call (when we expect it always to throw)

it's the second one that caused issues - see https://issues.apache.org/jira/browse/KAFKA-6994 - but that IMHO is unrelated to the changes in this PR.

if the second timeout is low - even without adding the required acl the test passed - that's the bug 6994.

if the second timeout is large the test takes longer, that's why we kept relatively short

ijuma replied:
Did you see my comment in KAFKA-6994?

edoardocomar replied:
yep - replied there - it's the same with the old `poll`

ijuma replied:
Even then, `poll` provides little in the way of guarantees. I suggest using `TestUtils.retry` with a generous timeout and a short timeout for `poll`. This also ensures we finish as quickly as possible.

edoardocomar replied:
good point - we can use `waitUntilTrue` checking for the topic being created as a consequence of `poll`

ijuma reviewed (commented):
A couple of quick questions while I review the rest.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -492,6 +495,26 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+
+    sendRequestAndVerifyResponseError(ApiKeys.CREATE_TOPICS, createTopicsRequest, resources, isAuthorized = false)
+
+    clusterCreateAcl.get(topicResource).foreach { acls =>
```
This is a bit dangerous because if `get` returns `None`, the test will pass.

edoardocomar replied:
Thanks! - the test did not actually need that block, (`get` was returning `None` !) was a sloppy remainder of previous refactoring

edoardocomar committed:
syntax simplification following review

edoardocomar committed:
addressed review comment

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1447,14 +1453,15 @@ class KafkaApis(val requestChannel: RequestChannel,
           } else Map.empty
 
-        val completeResults = results ++ duplicatedTopicsResults
+        val unauthorizedTopicsResults =  unauthorizedTopics.keySet.map(_ -> new ApiError(Errors.TOPIC_AUTHORIZATION_FAILED, null)) 
```
Extra space after `=`.

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -551,18 +566,34 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
-    val newTopic = "newTopic"
+  def testCreatePermissionOnTopicToWriteToNonExistentTopic() {
+    testCreatePermissionNeededToWriteToNonExistentTopic("newTopic",
+      Set(new Acl(userPrincipal, Allow, Acl.WildCardHost, Write), new Acl(userPrincipal, Allow, Acl.WildCardHost, Create)),
```
Why do we need `Write` here?

edoardocomar replied:
thanks, going to simplify by removing the constant arguments

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -160,6 +171,7 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
   this.serverConfig.setProperty(KafkaConfig.OffsetsTopicPartitionsProp, "1")
   this.serverConfig.setProperty(KafkaConfig.OffsetsTopicReplicationFactorProp, "3")
   this.serverConfig.setProperty(KafkaConfig.MinInSyncReplicasProp, "3")
+  this.serverConfig.setProperty(KafkaConfig.DefaultReplicationFactorProp, "3")
```
Why is this needed?

edoardocomar replied:
because we added tests which use topic auto-create, so they need a default replication factor matching the min in sync

edoardocomar replied:
thanks anyway, we spotted a missing `@Test` :(

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1034,19 +1035,20 @@ class KafkaApis(val requestChannel: RequestChannel,
         if (!authorize(request.session, Create, Resource.ClusterResource)) {
-          authorizedTopics --= nonExistingTopics
-          unauthorizedForCreateTopics ++= nonExistingTopics
+          val unauthorizedForCreateTopics = authorizedTopics.filter(topic => !authorize(request.session, Create, new Resource(Topic, topic)))
```
Nit: long line.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1034,19 +1035,20 @@ class KafkaApis(val requestChannel: RequestChannel,
       topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
 
-    var unauthorizedForCreateTopics = Set[String]()
+    var authorizedForDescribeNotCreateTopics = Set[String]()
```
Why did we rename this?

edoardocomar replied:
the new name though longer and not pretty matches the what the set of topics actually is in terms of authorizations

ijuma replied:
But then later we still have things like `unauthorizedForCreateTopicMetadata`. I think we should just keep this variable as it was since there is no actual change here. I'll push a minor commit.

edoardocomar committed:
using TestUtil loop to check for topic existence

edoardocomar committed:
fixed extra space

edoardocomar committed:
addressed review comment

edoardocomar committed:
missing @Test annotation

edoardocomar committed:
split long line

edoardocomar committed:
removed overkill testProduceConsumeTopicAutoCreateClusterCreateAcl

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1034,19 +1035,21 @@ class KafkaApis(val requestChannel: RequestChannel,
         if (!authorize(request.session, Create, Resource.ClusterResource)) {
-          authorizedTopics --= nonExistingTopics
-          unauthorizedForCreateTopics ++= nonExistingTopics
+          val unauthorizedForCreateTopics = authorizedTopics.filter(
```
This seems like a bug, we should only be doing this check for `nonExistingTopics`. Will push a fix.

ijuma committed:
Fix bug in KafkaApis

ijuma committed:
Minor tweaks in AuthorizerIntegrationTest

ijuma committed:
Remove dead code and some renames in EndToEndAuthorizationTest

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kip-277-fine-grained-acl-create-topics

* apache-github/trunk:
  KAFKA-5588: Remove deprecated --new-consumer tools option (#5097)
  MINOR: Fix for the location of the trogdor.sh executable file in the documentation. (#5040)
  KAFKA-6997: Exclude test-sources.jar when $INCLUDE_TEST_JARS is FALSE
  MINOR: docs should point to latest version (#5132)
  KAFKA-6981: Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes (KIP-298)
  [KAFKA-6730] Simplify State Store Recovery (#5013)
  MINOR: Rename package `internal` to `internals` for consistency (#5137)
  KAFKA-6704: InvalidStateStoreException from IQ when StreamThread closes store (#4801)
  MINOR: Add missing configs for resilience settings
  MINOR: Add regression tests for KTable mapValues and filter (#5134)
  KAFKA-6750: Add listener name to authentication context (KIP-282) (#4829)
  KAFKA-3665: Enable TLS hostname verification by default (KIP-294) (#4956)
  KAFKA-6938: Add documentation for accessing Headers on Kafka Streams Processor API (#5128)
  KAFKA-6813: return to double-counting for count topology names (#5075)
  KAFKA-5919; Adding checks on "version" field for tools using it
  MINOR: Remove deprecated KafkaStreams constructors in docs (#5118)

ijuma reviewed (approved):
I pushed a few minor changes including a merge conflict fix. This PR LGTM now and I will merge once the tests pass. @edoardocomar, please follow up if you have concerns about any of my changes.

omkreddy commented:
LGTM

edoardocomar commented:
thanks @ijuma

edoardocomar commented:
@ijuma we looked closely at your commit dc2727e
great that you spotted that we split the wrong collection ...
we added a minor PR to cover that case: #5155

ijuma commented:
@edoardocomar you read my mind as I was going to ask if you were willing to submit a PR to fix the testing gap. :)
